Fix rollback for missing resources

fmuyassarov opened the pull request:
**What this PR does / why we need it**:
Fix rollback for missing resources.
Fixes: https://github.com/helm/helm/issues/13437 

How to test:

Build the helm binary
```
$ make build
$ helm version
  version.BuildInfo{Version:"v4.0+unreleased", GitCommit:"f82e2b3c50d346948ce42e6a4f1348dee29aba31", GitTreeState:"dirty", GoVersion:"go1.24.7", KubeClientVersion:"v1.34"}
```
A quick test can be performed by creating a following charts directory:
```sh
├── v1
│   ├── Chart.yaml
│   ├── templates
│   │   ├── configmap.yaml
│   │   └── failure-hook.yaml
│   └── values.yaml
└── v2
    ├── Chart.yaml
    ├── templates
    │   └── failure-hook.yaml
    └── values.yaml
```

```YAML
{{- if .Release.IsInstall }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: cfg
data:
  foo: bar
{{- end }}
```
```YAML
{{- if .Release.IsUpgrade }}
apiVersion: batch/v1
kind: Job
metadata:
  name: failurecauser
  annotations:
    "helm.sh/hook": pre-upgrade
    "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded,hook-failed
spec:
  template:
    spec:
      terminationGracePeriodSeconds: 1
      containers:
        - name: never
          image: busybox:latest
          args: ["/bin/sh", "-c", "false"]
          imagePullPolicy: Always
      restartPolicy: Never
  backoffLimit: 0
{{- end }}
```

```shell
helm install test ./v1
NAME: test
LAST DEPLOYED: Wed Oct 22 12:34:41 2025
NAMESPACE: default
STATUS: deployed
REVISION: 1
DESCRIPTION: Install complete
TEST SUITE: None
```
```sh
helm upgrade test ./v2
^CRelease test has been cancelled.
level=WARN msg="upgrade failed" name=test error="context canceled"
Error: UPGRADE FAILED: context canceled
```
```sh
helm list -A
NAME	NAMESPACE	REVISION	UPDATED                              	STATUS	CHART              	APP VERSION
test	default  	2       	2025-10-22 12:35:55.186407 +0300 EEST	failed	helm-bug-demo-2.0.0	1.0
```
```sh
helm rollback test 1
Rollback was a success! Happy Helming!
```
```sh
helm list -A
NAME	NAMESPACE	REVISION	UPDATED                              	STATUS  	CHART              	APP VERSION
test	default  	3       	2025-10-22 12:37:00.757802 +0300 EEST	deployed	helm-bug-demo-1.0.0	1.0
```
**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

fmuyassarov committed:
Fix rollback for missing resources

Signed-off-by: Feruzjon Muyassarov <feruzjon.muyassarov@est.tech>

kashifest reviewed (approved):
Lgtm. It would be nice to have people with review rights to review this long standing fix

fmuyassarov commented:
> Hello @fmuyassarov
> 
> I’ve tested the patch and it works as expected. I’ve quickly tried to find what could go wrong with this new handling, and so far the patch looks correct. Maybe we’re missing a test that reflects what you’re fixing?
> Scenario for my test

Thanks for the review @benoittgt . 
I've addressed your comments. PTAL.

benoittgt commented:
@fmuyassarov it seems the push force didn't work. I don't see your [changes](https://github.com/helm/helm/compare/6c5fe2d91fd51ab2c528ca69eb48bec2d7cad8fe..cee7c0e30d0b42eb57428b3e05e39fd27549ec15).

fmuyassarov commented:
> @fmuyassarov it seems the push force didn't work. I don't see your [changes](https://github.com/helm/helm/compare/6c5fe2d91fd51ab2c528ca69eb48bec2d7cad8fe..cee7c0e30d0b42eb57428b3e05e39fd27549ec15).

Sorry @benoittgt , somehow missed those changes. Should be fixed now.

TerryHowe reviewed (dismissed):
/lgtm

fmuyassarov commented:
ping @mattfarina

fmuyassarov commented:
Hi @mattfarina, do you have updates on this?

fmuyassarov commented on pkg/kube/client.go:
```diff
@@ -531,14 +525,39 @@ func (c *Client) update(originals, targets ResourceList, updateApplyFunc UpdateA
+
+			currentObj, err := helper.Get(target.Namespace, target.Name)
+			if err != nil {
+				return fmt.Errorf("failed to get current resource state: %w", err)
```
Hi @TerryHowe, I can fix that. I just want to confirm whether a force push would remove the existing approval. If it’s not a big deal, I can submit a follow-up change; otherwise, I’ll fix it right away.

fmuyassarov replied:
Fixed.

TerryHowe reviewed (approved):
/lgtm
